File: windows/src/OPLFleetAgent.Core/TokenEventParser.cs
using System.Globalization;
using System.Text.Json;

namespace OPLFleetAgent.Core;

public sealed class TokenParserState
{
    internal string? CurrentModel { get; set; }
    internal string? Provider { get; set; }
    internal string? SessionIdFromMeta { get; set; }
    internal string? ForkParentId { get; set; }
    internal string? ChildSessionId { get; set; }
    internal string? ChildProvider { get; set; }
    internal string? ReplaySessionId { get; set; }
    internal bool WaitingForChildTurn { get; set; }
    internal bool ChildIsUserFork { get; set; }
    internal HashSet<string> ChildTaskStartedTurnIds { get; } = new(StringComparer.Ordinal);
    internal UsageTotals? PreviousTotals { get; set; }
    internal UsageTotals? InheritedBaseline { get; set; }
    internal long? InheritedReportedTotal { get; set; }
}

public static class TokenEventParser
{
    internal static readonly string[] RelevantMarkers =
    [
        "\"type\":\"token_count\",\"info\":",
        "\"type\":\"session_meta\",\"payload\":",
        "\"type\":\"turn_context\",\"payload\":",
        "\"type\":\"task_started\"",
    ];

    public static TokenParseBatch Parse(
        IEnumerable<string> lines,
        TokenParserState state,
        string fallbackSessionId)
    {
        var events = new List<UsageEvent>();
        var malformed = 0;

        foreach (var line in lines.Where(ShouldInspect))
        {
            try
            {
                using var document = JsonDocument.Parse(line);
                events.AddRange(Process(document.RootElement, state, fallbackSessionId));
            }
            catch (JsonException)
            {
                malformed++;
            }
        }

        return new TokenParseBatch(events, malformed);
    }

    internal static bool ShouldInspect(string line) =>
        RelevantMarkers.Any(marker => line.Contains(marker, StringComparison.Ordinal));

    private static IReadOnlyList<UsageEvent> Process(
        JsonElement entry,
        TokenParserState state,
        string fallbackSessionId)
    {
        var entryType = StringProperty(entry, "type");
        if (!ObjectProperty(entry, "payload", out var payload))
        {
            return [];
        }

        if (state.WaitingForChildTurn)
        {
            var turnId = StringProperty(payload, "turn_id");
            if (entryType == "turn_context" && ChildTurnStartsOwnSession(state, turnId))
            {
                state.WaitingForChildTurn = false;
                state.ReplaySessionId = null;
                state.ChildTaskStartedTurnIds.Clear();
                state.ChildIsUserFork = false;
                state.SessionIdFromMeta = state.ChildSessionId;
                state.Provider = NonEmpty(state.ChildProvider) ?? state.Provider;
                state.CurrentModel = ResolvedModel(payload) ?? state.CurrentModel;
            }
            else
            {
                RememberReplayState(entryType, payload, state);
                return [];
            }
        }

        if (entryType == "session_meta")
        {
            ProcessSessionMeta(payload, state);
            return [];
        }

        if (entryType == "turn_context")
        {
            state.CurrentModel = ResolvedModel(payload) ?? state.CurrentModel;
            return [];
        }

        if (entryType != "event_msg" || StringProperty(payload, "type") != "token_count" ||
            !ObjectProperty(payload, "info", out var info))
        {
            return [];
        }

        state.CurrentModel = ResolvedModel(payload) ?? state.CurrentModel;
        return ProcessTokenCount(entry, info, state, fallbackSessionId);
    }

    private static void ProcessSessionMeta(JsonElement payload, TokenParserState state)
    {
        var parentId = ResolvedForkParentId(payload);
        if (parentId is not null)
        {
            var id = NonEmpty(StringProperty(payload, "id"));
            var repeatedChild = !state.WaitingForChildTurn && state.ChildSessionId is not null &&
                state.ChildSessionId == id;
            var provider = NonEmpty(StringProperty(payload, "model_provider"));

            state.ForkParentId = parentId;
            state.ChildSessionId = id;
            state.ChildProvider = provider ?? state.ChildProvider;
            state.SessionIdFromMeta = id ?? state.SessionIdFromMeta;
            state.Provider = provider ?? state.Provider;

            if (!repeatedChild)
            {
                state.WaitingForChildTurn = true;
                state.ReplaySessionId = null;
                state.InheritedBaseline = null;
                state.InheritedReportedTotal = null;
                state.ChildTaskStartedTurnIds.Clear();
                state.ChildIsUserFork = StringProperty(payload, "thread_source") == "user";
            }
            return;
        }

        state.SessionIdFromMeta = NonEmpty(StringProperty(payload, "id")) ?? state.SessionIdFromMeta;
        state.Provider =
            NonEmpty(StringProperty(payload, "model_provider")) ?? state.Provider;
        state.CurrentModel = ResolvedModel(payload) ?? state.CurrentModel;
    }

    private static void RememberReplayState(
        string? entryType,
        JsonElement payload,
        TokenParserState state)
    {
        var payloadType = StringProperty(payload, "type");
        if (entryType == "event_msg" && payloadType == "task_started" &&
            NonEmpty(StringProperty(payload, "turn_id")) is { } turnId)
        {
            state.ChildTaskStartedTurnIds.Add(turnId);
        }

        if (entryType == "session_meta" && NonEmpty(StringProperty(payload, "id")) is { } id &&
            id != state.ChildSessionId)
        {
            state.ReplaySessionId = id;
        }

        if (entryType == "event_msg" && payloadType == "token_count" &&
            ObjectProperty(payload, "info", out var info) &&
            UsageProperty(info, "total_token_usage") is { } usage)
        {
            var totals = new UsageTotals(usage);
            state.PreviousTotals = totals;
            state.InheritedBaseline = totals;
            state.InheritedReportedTotal = usage.TotalTokens;
        }
    }

    private static IReadOnlyList<UsageEvent> ProcessTokenCount(
        JsonElement entry,
        JsonElement info,
        TokenParserState state,
        string fallbackSessionId)
    {
        UsageTotals? total = UsageProperty(info, "total_token_usage") is { } totalUsage
            ? new UsageTotals(totalUsage)
            : null;
        UsageTotals? last = UsageProperty(info, "last_token_usage") is { } lastUsage
            ? new UsageTotals(lastUsage)
            : null;

        if (ShouldSkipInherited(total, state))
        {
            return [];
        }
        state.InheritedBaseline = null;
        state.InheritedReportedTotal = null;

        TokenUsage usage;
        UsageTotals? nextTotals;
        if (total is { } current && last is { } increment && state.PreviousTotals is { } previous)
        {
            if (current == previous)
            {
                return [];
            }
            if (current.DeltaFrom(previous) is null &&
                current.LooksLikeStaleRegression(previous, increment))
            {
                return [];
            }
            usage = increment.AsUsage();
            nextTotals = current;
        }
        else if (total is { } initialTotal && last is { } initialIncrement)
        {
            usage = initialIncrement.AsUsage();
            nextTotals = initialTotal;
        }
        else if (total is { } cumulative && state.PreviousTotals is { } prior)
        {
            var delta = cumulative.DeltaFrom(prior);
            if (delta is null)
            {
                state.PreviousTotals = cumulative;
                return [];
            }
            usage = delta.Value.AsUsage();
            nextTotals = cumulative;
        }
        else if (total is { } firstTotal)
        {
            usage = firstTotal.AsUsage();
            nextTotals = firstTotal;
        }
        else if (last is { } onlyIncrement)
        {
            usage = onlyIncrement.AsUsage();
            nextTotals = state.PreviousTotals;
        }
        else
        {
            return [];
        }

        if (usage.TotalTokens <= 0)
        {
            return [];
        }
        state.PreviousTotals = nextTotals;

        var timestamp = ParseTimestamp(StringProperty(entry, "timestamp")) ?? DateTimeOffset.UtcNow;
        var sessionId = state.SessionIdFromMeta ?? fallbackSessionId;
        var scopeId = state.ForkParentId ?? sessionId;
        var provider = state.Provider ?? "unknown";
        var model = NonEmpty(StringProperty(info, "model")) ?? state.CurrentModel ?? "unknown";
        var key = DeduplicationKey(timestamp, usage, total, scopeId, provider, model);
        return [new UsageEvent(timestamp, usage, sessionId, key)];
    }

    private static bool ShouldSkipInherited(UsageTotals? total, TokenParserState state)
    {
        if (total is { } totals && state.InheritedReportedTotal is { } reported &&
            totals.ReportedTotal <= reported)
        {
            return true;
        }
        return total is { } current && state.InheritedBaseline is { } baseline &&
            current.IsWithin(baseline);
    }

    private static bool ChildTurnStartsOwnSession(TokenParserState state, string? turnId)
    {
        if (state.ReplaySessionId is null)
        {
            return true;
        }
        var childPrefix = UuidV7MillisecondPrefix(state.ChildSessionId);
        var turnPrefix = UuidV7MillisecondPrefix(turnId);
        if (childPrefix is null || turnPrefix is null)
        {
            return false;
        }
        var comparison = string.CompareOrdinal(turnPrefix, childPrefix);
        if (comparison > 0)
        {
            return true;
        }
        if (comparison < 0)
        {
            return false;
        }
        return state.ChildIsUserFork || state.ChildTaskStartedTurnIds.Contains(turnId!);
    }

    private static string? UuidV7MillisecondPrefix(string? id)
    {
        if (id is null)
        {
            return null;
        }
        var parts = id.Split('-');
        if (parts.Length != 5 || parts[0].Length != 8 || parts[1].Length != 4 ||
            parts[2].Length != 4 || parts[2][0] != '7')
        {
            return null;
        }
        var prefix = (parts[0] + parts[1]).ToLowerInvariant();
        return prefix.All(Uri.IsHexDigit) ? prefix : null;
    }

    private static string DeduplicationKey(
        DateTimeOffset timestamp,
        TokenUsage usage,
        UsageTotals? total,
        string scopeId,
        string provider,
        string model)
    {
        var time = timestamp.ToUniversalTime().ToString("O", CultureInfo.InvariantCulture);
        return total is { } totals
            ? string.Join(':', "codex", "total", scopeId, provider, model, time,
                totals.Input, totals.Output, totals.Cached, totals.Reasoning,
                totals.ReportedTotal)
            : string.Join(':', "codex", "event", scopeId, provider, model, time,
                usage.InputTokens, usage.CachedInputTokens, usage.OutputTokens,
                usage.ReasoningOutputTokens);
    }

    private static DateTimeOffset? ParseTimestamp(string? value) =>
        DateTimeOffset.TryParse(
            value,
            CultureInfo.InvariantCulture,
            DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal,
            out var date)
            ? date
            : null;

    private static TokenUsage? UsageProperty(JsonElement parent, string name)
    {
        if (!ObjectProperty(parent, name, out var usage))
        {
            return null;
        }
        return new TokenUsage(
            LongProperty(usage, "input_tokens"),
            LongProperty(usage, "cached_input_tokens"),
            LongProperty(usage, "output_tokens"),
            LongProperty(usage, "reasoning_output_tokens"),
            NullableLongProperty(usage, "total_tokens"));
    }

    private static string? ResolvedModel(JsonElement payload)
    {
        string? slug = null;
        if (ObjectProperty(payload, "model_info", out var modelInfo))
        {
            slug = NonEmpty(StringProperty(modelInfo, "slug"));
        }
        string? infoModel = null;
        if (ObjectProperty(payload, "info", out var info))
        {
            infoModel = NonEmpty(StringProperty(info, "model"));
        }
        return slug ?? NonEmpty(StringProperty(payload, "model")) ??
            NonEmpty(StringProperty(payload, "model_name")) ?? infoModel;
    }

    private static string? ResolvedForkParentId(JsonElement payload)
    {
        var direct = NonEmpty(StringProperty(payload, "forked_from_id"));
        if (direct is not null)
        {
            return direct;
        }
        if (!ObjectProperty(payload, "source", out var source) ||
            !ObjectProperty(source, "subagent", out var subagent) ||
            !ObjectProperty(subagent, "thread_spawn", out var spawn))
        {
            return null;
        }
        return NonEmpty(StringProperty(spawn, "parent_thread_id"));
    }

    private static bool ObjectProperty(JsonElement parent, string name, out JsonElement value)
    {
        if (parent.ValueKind == JsonValueKind.Object && parent.TryGetProperty(name, out value) &&
            value.ValueKind == JsonValueKind.Object)
        {
            return true;
        }
        value = default;
        return false;
    }

    private static string? StringProperty(JsonElement parent, string name) =>
        parent.ValueKind == JsonValueKind.Object && parent.TryGetProperty(name, out var value) &&
        value.ValueKind == JsonValueKind.String
            ? value.GetString()
            : null;

    private static long LongProperty(JsonElement parent, string name) =>
        NullableLongProperty(parent, name) ?? 0;

    private static long? NullableLongProperty(JsonElement parent, string name) =>
        parent.ValueKind == JsonValueKind.Object && parent.TryGetProperty(name, out var value) &&
        value.TryGetInt64(out var number)
            ? number
            : null;

    private static string? NonEmpty(string? value) =>
        string.IsNullOrWhiteSpace(value) ? null : value.Trim();
}

File: Sources/OPLFleetAgentCore/TokenEventParser.swift
import Foundation

public struct TokenParserState: Sendable {
  var currentModel: String?
  var provider: String?
  var sessionIDFromMeta: String?
  var forkParentID: String?
  var childSessionID: String?
  var childProvider: String?
  var replaySessionID: String?
  var waitingForChildTurn = false
  var childIsUserFork = false
  var childTaskStartedTurnIDs: Set<String> = []
  var previousTotals: UsageTotals?
  var inheritedBaseline: UsageTotals?
  var inheritedReportedTotal: Int64?

  public init() {}
}

public struct TokenParseBatch: Sendable {
  public let events: [UsageEvent]
  public let malformedRelevantLines: Int

  public init(events: [UsageEvent], malformedRelevantLines: Int) {
    self.events = events
    self.malformedRelevantLines = malformedRelevantLines
  }
}

public enum TokenEventParser {
  static let relevantMarkers = [
    Data("\"type\":\"token_count\",\"info\":".utf8),
    Data("\"type\":\"session_meta\",\"payload\":".utf8),
    Data("\"type\":\"turn_context\",\"payload\":".utf8),
    Data("\"type\":\"task_started\"".utf8),
  ]

  public static func parse(
    lines: [Data],
    state: inout TokenParserState,
    fallbackSessionID: String
  ) -> TokenParseBatch {
    var events: [UsageEvent] = []
    var malformed = 0
    let decoder = JSONDecoder()

    for line in lines where shouldInspect(line) {
      let entry: CodexLogEntry
      do {
        entry = try decoder.decode(CodexLogEntry.self, from: line)
      } catch {
        malformed += 1
        continue
      }

      events.append(
        contentsOf: process(
          entry,
          state: &state,
          fallbackSessionID: fallbackSessionID
        ))
    }

    return TokenParseBatch(events: events, malformedRelevantLines: malformed)
  }

  static func shouldInspect(_ line: Data) -> Bool {
    relevantMarkers.contains { line.range(of: $0) != nil }
  }

  private static func process(
    _ entry: CodexLogEntry,
    state: inout TokenParserState,
    fallbackSessionID: String
  ) -> [UsageEvent] {
    guard let payload = entry.payload else { return [] }

    if state.waitingForChildTurn {
      if entry.type == "turn_context",
        childTurnStartsOwnSession(state: state, turnID: payload.turnID)
      {
        state.waitingForChildTurn = false
        state.replaySessionID = nil
        state.childTaskStartedTurnIDs.removeAll()
        state.childIsUserFork = false
        state.sessionIDFromMeta = state.childSessionID
        state.provider = state.childProvider ?? state.provider
        state.currentModel = payload.resolvedModel ?? state.currentModel
      } else {
        rememberReplayState(entry, payload: payload, state: &state)
        return []
      }
    }

    if entry.type == "session_meta" {
      processSessionMeta(payload, state: &state)
      return []
    }

    if entry.type == "turn_context" {
      state.currentModel = payload.resolvedModel ?? state.currentModel
      return []
    }

    guard entry.type == "event_msg", payload.type == "token_count", let info = payload.info else {
      return []
    }

    state.currentModel = payload.resolvedModel ?? state.currentModel
    return processTokenCount(
      entry,
      info: info,
      state: &state,
      fallbackSessionID: fallbackSessionID
    )
  }

  private static func processSessionMeta(_ payload: CodexPayload, state: inout TokenParserState) {
    if let parentID = payload.resolvedForkParentID {
      let repeatedChild =
        !state.waitingForChildTurn
        && state.childSessionID != nil
        && state.childSessionID == payload.id

      state.forkParentID = parentID
      state.childSessionID = payload.id
      state.childProvider = payload.modelProvider.nonEmpty ?? state.childProvider
      state.sessionIDFromMeta = payload.id ?? state.sessionIDFromMeta
      state.provider = payload.modelProvider.nonEmpty ?? state.provider

      if !repeatedChild {
        state.waitingForChildTurn = true
        state.replaySessionID = nil
        state.inheritedBaseline = nil
        state.inheritedReportedTotal = nil
        state.childTaskStartedTurnIDs.removeAll()
        state.childIsUserFork = payload.threadSource == "user"
      }
      return
    }

    state.sessionIDFromMeta = payload.id ?? state.sessionIDFromMeta
    state.provider = payload.modelProvider.nonEmpty ?? state.provider
    state.currentModel = payload.resolvedModel ?? state.currentModel
  }

  private static func rememberReplayState(
    _ entry: CodexLogEntry,
    payload: CodexPayload,
    state: inout TokenParserState
  ) {
    if entry.type == "event_msg", payload.type == "task_started",
      let turnID = payload.turnID.nonEmpty
    {
      state.childTaskStartedTurnIDs.insert(turnID)
    }

    if entry.type == "session_meta",
      let id = payload.id.nonEmpty,
      id != state.childSessionID
    {
      state.replaySessionID = id
    }

    if entry.type == "event_msg",
      payload.type == "token_count",
      let totalUsage = payload.info?.totalTokenUsage
    {
      let totals = UsageTotals(totalUsage)
      state.previousTotals = totals
      state.inheritedBaseline = totals
      state.inheritedReportedTotal = totalUsage.totalTokens
    }
  }

  private static func processTokenCount(
    _ entry: CodexLogEntry,
    info: CodexUsageInfo,
    state: inout TokenParserState,
    fallbackSessionID: String
  ) -> [UsageEvent] {
    let total = info.totalTokenUsage.map(UsageTotals.init)
    let last = info.lastTokenUsage.map(UsageTotals.init)

    if shouldSkipInherited(total: total, state: state) {
      return []
    }
    state.inheritedBaseline = nil
    state.inheritedReportedTotal = nil

    let usage: TokenUsage
    let nextTotals: UsageTotals?

    switch (total, last, state.previousTotals) {
    case (.some(let current), .some(let increment), .some(let previous)):
      if current == previous {
        return []
      }
      if current.delta(from: previous) == nil,
        current.looksLikeStaleRegression(previous: previous, last: increment)
      {
        return []
      }
      usage = increment.asUsage()
      nextTotals = current

    case (.some(let current), .some(let increment), .none):
      usage = increment.asUsage()
      nextTotals = current

    case (.some(let current), .none, .some(let previous)):
      guard let delta = current.delta(from: previous) else {
        state.previousTotals = current
        return []
      }
      usage = delta.asUsage()
      nextTotals = current

    case (.some(let current), .none, .none):
      usage = current.asUsage()
      nextTotals = current

    case (.none, .some(let increment), _):
      usage = increment.asUsage()
      nextTotals = state.previousTotals

    case (.none, .none, _):
      return []
    }

    guard usage.totalTokens > 0 else { return [] }
    state.previousTotals = nextTotals

    let timestamp = parseTimestamp(entry.timestamp) ?? Date()
    let sessionID = state.sessionIDFromMeta ?? fallbackSessionID
    let scopeID = state.forkParentID ?? sessionID
    let provider = state.provider ?? "unknown"
    let model = info.model.nonEmpty ?? state.currentModel ?? "unknown"
    let key = deduplicationKey(
      timestamp: timestamp,
      usage: usage,
      total: total,
      scopeID: scopeID,
      provider: provider,
      model: model
    )

    return [
      UsageEvent(
        timestamp: timestamp,
        usage: usage,
        sessionID: sessionID,
        deduplicationKey: key
      )
    ]
  }

  private static func shouldSkipInherited(total: UsageTotals?, state: TokenParserState) -> Bool {
    if let reported = total?.reportedTotal,
      let baseline = state.inheritedReportedTotal,
      reported <= baseline
    {
      return true
    }

    if let total, let baseline = state.inheritedBaseline {
      return total.isWithin(baseline)
    }
    return false
  }

  private static func childTurnStartsOwnSession(state: TokenParserState, turnID: String?) -> Bool {
    guard state.replaySessionID != nil else { return true }
    guard
      let childID = state.childSessionID,
      let turnID,
      let childPrefix = uuidV7MillisecondPrefix(childID),
      let turnPrefix = uuidV7MillisecondPrefix(turnID)
    else {
      return false
    }

    if turnPrefix > childPrefix { return true }
    if turnPrefix < childPrefix { return false }
    return state.childIsUserFork || state.childTaskStartedTurnIDs.contains(turnID)
  }

  private static func uuidV7MillisecondPrefix(_ id: String) -> String? {
    let parts = id.split(separator: "-")
    guard parts.count == 5,
      parts[0].count == 8,
      parts[1].count == 4,
      parts[2].count == 4,
      parts[2].first == "7"
    else {
      return nil
    }
    let prefix = String(parts[0] + parts[1]).lowercased()
    guard prefix.allSatisfy({ $0.isHexDigit }) else { return nil }
    return prefix
  }

  private static func parseTimestamp(_ value: String?) -> Date? {
    guard let value else { return nil }
    return try? Date(value, strategy: .iso8601)
  }

  private static func deduplicationKey(
    timestamp: Date,
    usage: TokenUsage,
    total: UsageTotals?,
    scopeID: String,
    provider: String,
    model: String
  ) -> String {
    if let total {
      return [
        "codex", "total", scopeID, provider, model, timestamp.formatted(.iso8601),
        String(total.input), String(total.output), String(total.cached),
        String(total.reasoning), String(total.reportedTotal),
      ].joined(separator: ":")
    }

    return [
      "codex", "event", scopeID, provider, model, timestamp.formatted(.iso8601),
      String(usage.inputTokens), String(usage.cachedInputTokens),
      String(usage.outputTokens), String(usage.reasoningOutputTokens),
    ].joined(separator: ":")
  }
}
